Add event feature

larrywang0701 opened the pull request:
- Implement commands `add event`, `list events` and `delete event`
   - Format (also made the updates of command format to User Guide):
     - `add event -id CONTACT_ID -en EVENT_NAME -st START_TIME [-et END_TIME] [-loc LOCATION] [-info INFORMATION]`
     - `list events`
     - `delete event -id CONTACT_ID -eid EVENT_ID`
   - Try it with (execute these commands one by one by sequence):
     - `add event -id 1 -en Meeting with professor -st 12:00 -et 01:00 -loc COM 1 Basement -info Discuss the project`
     - `add event -id 9999 -en Meeting with professor -st 12:00 -et 01:00 -loc COM 1 Basement -info Discuss the project` **(Error: Can not find target contact)**
     - `list events`
     - `delete event -id 1 -eid 1`
     - `delete event -id 999 -eid 1` **(Error: Can not find target contact)**
     - `delete event -id 1 -eid 999` **(Error: event not found)**
  
    - For `add event`: now events with **duplicate event names** are now **allowed**

- Adjust command parser to support secondary command word (e.g. The `event` part in `add event`.)
  - The original code in <i>Add/List/DeleteCommand(Parser).java</i> are moved to <i>Add/List/Delete**Person**Command(Parser).java</i> and assigned the secondary command `contact` (e.g. `add contact`, `delete contact`), other stuff remain unchanged

- Fix all CheckStyle errors in the project

- Adjust JUnit Test Cases to let it work on our secondary command formats for `add/list/delete contact ...`

 - Add JUnit test cases for
   - `AddEventCommand`
   - `ListEventCommand`
   - `DeleteEventCommand`
   - `AddCommandParser` (also tests `AddEventCommandParser`)
   - `DeleteCommandParser` (also tests `DeleteEventCommandParser`)
   - `ListCommandParser`
   - `DateTimeUtil`

larrywang0701 committed:
Add event feature
Implement commands `add event`, `list events` and `delete event`
Adjust command parser to support secondary command word
(e.g. The `event` part in `add event`.)
The original code in Add/List/DeleteCommand(Parser) are moved to
Add/List/Delete**Person**Command(Parser) and assigned the secondary command
`contact` (e.g. `add contact`, `delete contact`), other stuff remain unchanged

larrywang0701 committed:
Fix code style errors

larrywang0701 committed:
Fix code style errors

larrywang0701 committed:
Fix code style errors

larrywang0701 committed:
Fix Compile Errors in JUnit Test Cases

larrywang0701 committed:
Fix Compile Error in JUnit Test Cases

larrywang0701 committed:
Fix compile error due to renaming variable name for code style

larrywang0701 committed:
Fix code style errors

larrywang0701 committed:
Fix checksyle errors in junit test case

larrywang0701 committed:
Adjust JUnit Test Cases to let it work on secondary commands `add/list/delete contact ...`

larrywang0701 committed:
Fix checkstyle errors in JUnit test codes

larrywang0701 committed:
Fix checkstyle error in JUnit test code

larrywang0701 committed:
Add JUnit Test Cases

Add JUnit test cases for
 - `AddEventCommand`
 - `ListEventCommand`
 - `DeleteEventCommand`
 - `AddCommandParser` (also tests `AddEventCommandParser`)
 - `DeleteCommandParser` (also tests `DeleteEventCommandParser`)
 - `ListCommandParser`
 - `DateTimeUtil`

larrywang0701 committed:
Fix checkstyle error

larrywang0701 committed:
Fix checkstyle error

larrywang0701 committed:
Fix checkstyle error

larrywang0701 committed:
Adjust test case for `ListEventCommand`

larrywang0701 committed:
Adjust test case for `ListEventCommand`

larrywang0701 committed:
Adjust test case for `ListEventCommand`

larrywang0701 committed:
Fix checkstyle error

larrywang0701 committed:
Fix checkstyle error

zekone commented on src/main/java/seedu/address/logic/commands/AddEventCommand.java:
```diff
@@ -0,0 +1,55 @@
+     * Creates an AddEventCommand to add the specified {@code Event}
+     */
+    public AddEventCommand(String contactName, Event event) {
+        requireNonNull(contactName);
+        requireNonNull(event);
```
Can consider using `requireAllNonNull(contactName, event)` instead?

zekone commented on src/main/java/seedu/address/logic/parser/AddressBookParser.java:
```diff
@@ -83,4 +84,12 @@ public Command parseCommand(String userInput) throws ParseException {
         }
     }
 
+    /**
+     * Returns true if none of the prefixes contains empty {@code Optional} values
+     * in the given
+     * {@code ArgumentMultimap}.
+     */
+    public static boolean arePrefixesPresent(ArgumentMultimap argumentMultimap, Prefix... prefixes) {
+        return Stream.of(prefixes).allMatch(prefix -> argumentMultimap.getValue(prefix).isPresent());
+    }
```
Maybe move to `ParserUtil`? I think it fits there better, OOP wise.

zekone commented on src/main/java/seedu/address/model/event/EventTime.java:
Should we allow `EventTime` to be `null`? I think event needs to have a time to be useful, without a start time, we cannot keep track of it at all. Furthermore, we are also planning to sort event in chronological order, allowing time to be `null` can be a problem later on.

Nixx162 replied:
Second this

larrywang0701 replied:
Because event end time is optional to exist, but the `Storage` requires all fields to be not null so it can be represented in JSON string. so I make the content (the `time` field) in EventTime nullable (not EventTime itself). Regarding your concern about <i>what will happen when an event is without a start time</i>, since the `add event` command will check whether `-st` parameter (start time) exists, thus **there shouldn't be any events without a valid start time**. If you still have concerns (that you still think `start time` can be null, although I think this is impossible except something is bugged), then the nullity check on start time `EventTime` should be done on the sorting process.

zekone commented on src/main/java/seedu/address/commons/util/DateTimeUtil.java:
```diff
@@ -0,0 +1,49 @@
+    public static LocalDateTime parseString(String str) throws DateTimeParseException {
+        DateTimeFormatter formatter1 = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm[:ss]");
+        LocalDateTime result = null;
+        try {
+            result = LocalDateTime.parse(str);
+        } catch (DateTimeParseException e) {
+            try {
+                result = LocalDateTime.parse(str, formatter1);
+            } catch (DateTimeParseException e2) {
+                LocalDateTime now = LocalDateTime.now();
+                String appendDate = now.getYear() + "-" + now.getMonthValue() + "-" + now.getDayOfMonth() + " ";
+                try {
+                    result = LocalDateTime.parse(appendDate + str, formatter1);
+                } catch (DateTimeParseException e3) {
+                    String appendZero = "0";
+                    result = LocalDateTime.parse(appendDate + appendZero + str, formatter1);
+                }
```
Seems too nested? Also, I am not sure what's happening in the last two try-catch clauses.

Nixx162 replied:
I really don't think this amount of try-catch is necessary. Especially the part where we're assuming that if the string format is messed up, then the return would just be the current time. We should simply educate the users through the user guide and make them stick to one or two formats, otherwise we just throw a formatting exception

zekone reviewed (commented):
Maybe we just use index instead of the name of the contact for the commands?

Nixx162 commented on src/main/java/seedu/address/logic/commands/AddEventCommand.java:
```diff
@@ -0,0 +1,55 @@
+        if (person == null) {
+            throw new CommandException(MESSAGE_PERSON_NOT_FOUNT + this.contactName);
+        }
+        boolean eventExists = person.getEvents().stream().anyMatch(e -> e.getName().equals(this.toAdd.getName()));
```
SLAP

Nixx162 replied:
Just make a callable function in the Person class to do this

larrywang0701 committed:
Use contact ID rather than contact names in event commands

 - Use contact ID rather than contact names in event commands
 - Adjust code as the PR comments

larrywang0701 commented:
Already changed. Read the PR information again to see the changes.

larrywang0701 committed:
Fix date-time parse error in sample data

larrywang0701 committed:
Fix an error in the example command in UG

larrywang0701 committed:
Fix one more error in UG

larrywang0701 committed:
Adjust the information about `CONTACT_ID` in UG

Nixx162 commented on src/main/java/seedu/address/logic/commands/AddEventCommand.java:
```diff
@@ -0,0 +1,49 @@
+        requireNonNull(model);
+        Person person = model.findPersonByUserFriendlyId(this.contactId);
+        if (person == null) {
+            throw new CommandException(MESSAGE_PERSON_NOT_FOUNT + this.contactId);
```
Typo

Nixx162 commented on src/main/java/seedu/address/logic/commands/AddEventCommand.java:
```diff
@@ -0,0 +1,49 @@
+            + ": Adds an event to a contact.\n"
+            + "Usage:  add event -n CONTACT_NAME -en EVENT_NAME -st "
+            + "START_TIME [-et END_TIME] [-loc LOCATION] [-i INFORMATION]";
+    public static final String MESSAGE_PERSON_NOT_FOUNT = "Can not find the target contact with ID: ";
```
Typo as well

Nixx162 commented on src/main/java/seedu/address/logic/commands/ListEventCommand.java:
```diff
@@ -0,0 +1,32 @@
+        requireNonNull(model);
+        List<Person> persons = model.getAddressBook().getPersonList();
+        StringBuilder str = new StringBuilder(MESSAGE);
+        persons.forEach(
+                person -> person.getEvents().forEach(
+                        event -> str.append("[").append(
+                                person.getName().toString()).append("] ")
+                                .append(event.getUiText()).append("\n")
+                )
+        );
```
Can we abstract this?

zekone commented on src/main/java/seedu/address/ui/MainWindow.java:
```diff
@@ -193,4 +198,16 @@ private CommandResult executeCommand(String commandText) throws CommandException
             throw e;
         }
     }
+
+    /**
+     * Refresh the person list panel and all person cards immediately
+     */
+    public static void refreshPersonListPanelImmediately() {
+        if (iNSTANCE == null) {
+            return;
+        }
+        iNSTANCE.personListPanelPlaceholder.getChildren().clear();
+        iNSTANCE.personListPanel = new PersonListPanel(iNSTANCE.logic.getFilteredPersonList());
+        iNSTANCE.personListPanelPlaceholder.getChildren().add(iNSTANCE.personListPanel.getRoot());
+    }
```
Is this really necessary? If we can add and delete contact previously, then adding and deleting events should not require this extra method. Also, it may not be a huge issue, but it is also breaking abstraction by allowing command to directly interact with UI instead of using the model that was passed to it. Perhaps we can try using `EditCommand` to change the event list or `Model` to `setPerson`

Nixx162 commented on src/main/java/seedu/address/commons/util/DateTimeUtil.java:
```diff
@@ -0,0 +1,40 @@
+        LocalDateTime result = null;
+        try {
+            result = LocalDateTime.parse(str, formatter1);
+        } catch (DateTimeParseException e2) {
+            LocalDateTime now = LocalDateTime.now();
+            String appendDate = now.getYear() + "-" + now.getMonthValue() + "-" + now.getDayOfMonth() + " ";
+            result = LocalDateTime.parse(appendDate + str, formatter1);
+        }
```
I think should be more reasonable to take care of the case where the user only puts in date (since this is a more general information) instead of the case where the user only puts in time. 

In case you wanna handle both, can you find a way to mkae sure that the code won't be too nested?

Nixx162 replied:
You can resolve this one without changing the code, but for the other conversations, can you please make the change? Thanks!

zekone reviewed (commented):
Everything seems to be working as intended but there may be some issues with implementation.

larrywang0701 committed:
Adjust code based on PR comments

larrywang0701 committed:
Make the event list UI refresh without breaking abstraction

larrywang0701 committed:
Add event id in UI and remove dead code

Nixx162 reviewed (approved):
LGTM

AndrewJanong commented:
LGTM
